### ISSUE TITLE
git module: always returns status changed

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#30113]
##### ISSUE TYPE

bug report

##### COMPONENT NAME

git module

##### ANSIBLE VERSION

1.9.1

##### OS / ENVIRONMENT

OS X 10.3 provisioning Ubuntu 14.04

##### SUMMARY

:**

git module always returns status changed, even if there was no changes at all.

The repo is cloned into `/opt`, which requires root permissions, however the repo itself cannot be cloned by the root user, so is cloned to `/tmp`, then moved into `/opt`. There's a second job to pull future updates as the `foo` user when the repo already exists.

```
- name: Check if scripts repository is already installed.
  stat: "path=/opt/scripts"
  register: scripts_repo

- name: Clone the scripts repository.
  git:
    repo: ssh://git@repo.example.com:/example.com/server-scripts.git
    dest: /tmp/scripts
    accept_hostkey: true
  sudo: yes
  sudo_user: foo
  when: not scripts_repo.stat.exists

- name: Move the scripts repo.
  command: mv /tmp/scripts /opt/scripts
  sudo: yes
  when: not scripts_repo.stat.exists

- name: Ensure scripts is up to date (when not cloning).
  git:
    repo: ssh://git@repo.example.com:/example.com/server-scripts.git
    dest: /opt/scripts
    accept_hostkey: true
  sudo: yes
  sudo_user: foo
  when: scripts_repo.stat.exists
  register: scripts_update 

- debug: var=scripts_update 
```

Output from task runner indicated that the before and after commit refs are the same, however changed is `true`:

```
TASK: [Check if scripts repository is already installed.] *********************
ok: [192.168.33.10]

TASK: [Clone the scripts repository.] *****************************************
skipping: [192.168.33.10]

TASK: [Move the scripts repo.] ************************************************
skipping: [192.168.33.10]

TASK: [Ensure scripts is up to date (when not cloning).] **********************
changed: [192.168.33.10]

TASK: [debug var=scripts_update] **********************************************
ok: [192.168.33.10] => {
    "var": {
        "scripts_update": {
            "after": "a8f34d434fcc0a1356aa4d8dda1eb1112bbff105",
            "before": "a8f34d434fcc0a1356aa4d8dda1eb1112bbff105",
            "changed": true,
            "invocation": {
                "module_args": "",
                "module_name": "git"
            }
        }
    }
}
```

I've found a [similar issue from 2014](https://github.com/ansible/ansible/issues/6526), which was caused by an unclean repository, however in this case the repo itself appears to be clean:

```
/opt/scripts $ sudo -u foo git status
On branch master
Your branch is up-to-date with 'origin/master'.

nothing to commit, working directory clean
```

**

##### STEPS TO REPRODUCE

:**

Unclear. We're doing the same round the houses approach for another repo, and that is reporting `changed` correctly.

**

##### EXPECTED RESULTS

:**

`scripts_update.changed` should be `false` when the before and after commit refs are the same.

**

##### ACTUAL RESULTS

:**

`scripts_update.changed` is `true`.

Copied from original issue: ansible/ansible-modules-core#1509

**Comment by ansibot** [comment]
From @philipforget on 2015-06-12T15:05:39Z

I'm seeing this behavior here locally, will try and pinpoint why the hashes are the same but changed is being set to `true`

**Comment by ansibot** [comment]
From @JohanTan on 2015-06-12T15:05:39Z

Confirmed on version 1.9.0.1. The parameter I used was:

```
connection: local
git: repo=ssh://git@github.com/foo/bar.git dest=/var/www/git_repos/bar update=yes track  _submodules=yes clone=yes
```

and it always says "changed: yes" even thought the SHA1 stays unchanged. But when I removed `track_submodules=yes` it seems to work properly. There might be something with that option.

**Comment by ansibot** [comment]
From @mmcdole on 2015-06-12T15:05:39Z

+1 also ran into this issue today.

**Comment by ansibot** [comment]
From @jmcvetta on 2015-06-12T15:05:39Z

Seeing this issue in 2.0.1.0 as well.  

I used `changed_when` as a workaround.  Result varaibles `before` and `after` contain Git commit IDs.

``` yaml
- name: Ensure application repo is checked out
  git:
    repo: "{{ repo }}"
    dest: "{{ dest }}"
    version: "{{ version }}"
  register: git_result
  changed_when: "git_result.after|default('after') != git_result.before|default('before')"
```

**Comment by ansibot** [comment]
From @tommarshall on 2015-06-12T15:05:39Z

Thanks for the workaround @jmcvetta. That worked for me in 1.9.4.

Just a word of warning, that workaround could cause issues if the `git` module's `force` option is set to `true`/`yes` and there are local file modifications. `git` would normally report changed in that situation, but won't with this workaround in place.

**Comment by ansibot** [comment]
From @nirik on 2015-06-12T15:05:39Z

Seeing this here also with: 

ansible-playbook 2.1.0.0
  config file = /home/kevin/.ansible.cfg
  configured module search path = Default w/o overrides

Test playbook: 

```
- name: test playbook for git bug
  hosts: thulsadoom.scrye.com
  gather_facts: False

  tasks:

  - name: git checkout test
    git:
      repo: https://github.com/ansible/ansible-modules-core.git
      dest: '/tmp/ansible-modules-core'
```

```
Results: 
 % ansible-playbook -u root -i inventory playbooks/test.yml

PLAY [test playbook for git bug] ***********************************************

TASK [git checkout test] *******************************************************
changed: [thulsadoom.scrye.com]

PLAY RECAP *********************************************************************
thulsadoom.scrye.com       : ok=1    changed=1    unreachable=0    failed=0   

(as expected, the first run shows changed because it cloned the repo).

% ansible-playbook -u root -i inventory playbooks/test.yml

PLAY [test playbook for git bug] ***********************************************

TASK [git checkout test] *******************************************************
ok: [thulsadoom.scrye.com]

PLAY RECAP *********************************************************************
thulsadoom.scrye.com       : ok=1    changed=0    unreachable=0    failed=0   

% ansible-playbook -u root -i inventory playbooks/test.yml

PLAY [test playbook for git bug] ***********************************************

TASK [git checkout test] *******************************************************
ok: [thulsadoom.scrye.com]

PLAY RECAP *********************************************************************
thulsadoom.scrye.com       : ok=1    changed=0    unreachable=0    failed=0   

(several normal runs in a row show no changes)

% ansible-playbook -u root -i inventory playbooks/test.yml --check

PLAY [test playbook for git bug] ***********************************************

TASK [git checkout test] *******************************************************
changed: [thulsadoom.scrye.com]

PLAY RECAP *********************************************************************
thulsadoom.scrye.com       : ok=1    changed=1    unreachable=0    failed=0   
```

(run with check shows changed, when it shouldn't)

verbose output from the check run: 

```
changed: [thulsadoom.scrye.com] => {"after": "f5e8204dbc0a3573a12a119413c00134d0ecf80b", "before": "f5e8204dbc0a3573a12a119413c00134d0ecf80b", "changed": true, "invocation": {"module_args": {"accept_hostkey": false, "bare": false, "clone": true, "depth": null, "dest": "/tmp/ansible-modules-core", "executable": null, "force": false, "key_file": null, "recursive": true, "reference": null, "refspec": null, "remote": "origin", "repo": "https://github.com/ansible/ansible-modules-core.git", "ssh_opts": null, "track_submodules": false, "update": true, "verify_commit": false, "version": "HEAD"}, "module_name": "git"}}
```

**Comment by ansibot** [comment]
From @willthames on 2015-06-12T15:05:39Z

@nirik, it's helpful if you wrap your output and playbook in triple backticks.

**Comment by ansibot** [comment]
From @nirik on 2015-06-12T15:05:39Z

Sorry. Edited, hopefully thats more readable now...

**Comment by ansibot** [comment]
From @willthames on 2015-06-12T15:05:39Z

@nirik, ok, this is a problem with checkmode only. 

It's fixed in #3025. I would expect that to go out with 2.1.1 but it doesn't appear to be in the RC, so I'll let @jimi-c know.

**Comment by ansibot** [comment]
From @nirik on 2015-06-12T15:05:39Z

Thanks much for looking at it.

**Comment by ansibot** [comment]
From @willthames on 2015-06-12T15:05:39Z

2.1.1 RC2 does indeed contain this commit.

**Comment by ansibot** [comment]
From @jctanner on 2015-06-12T15:05:39Z

Testcase ...

```
- hosts: el7host
  user: vagrant
  gather_facts: False
  vars:
    checkoutdir: /tmp/amctest
    checkoutdir2: /root/amctest
  tasks:
    - name: clean checkout
      file:
        path: "{{ checkoutdir }}"
        state: absent
    - name: clean checkout2
      file:
        path: "{{ checkoutdir2 }}"
        state: absent
      sudo: yes

    - name: make initial clone
      git:
        repo: https://github.com/ansible/ansible-modules-core
        dest: "{{ checkoutdir }}"
        version: devel

    ######################################
    # TEST CASE
    ######################################

    - name: move the checkout
      shell: mv "{{ checkoutdir }}" "{{ checkoutdir2 }}"
      sudo: yes

    - name: re-clone
      git:
        repo: https://github.com/ansible/ansible-modules-core
        dest: "{{ checkoutdir2 }}"
        version: devel
      sudo: yes
      register: recheckout

    - debug: var=recheckout
    - assert:
        that:
            - "not recheckout.changed"
```

Results ...

```
...
v1.6.10 ; 0
v1.6.2 ; 0
v1.6.3 ; 0
v1.6.4 ; 0
v1.6.5 ; 0
v1.6.6 ; 0
v1.6.7 ; 0
v1.6.8 ; 0
v1.6.9 ; 0
v1.7.0 ; 0
v1.7.1 ; 0
v1.7.2 ; 0
v1.8.0 ; 0
v1.8.1 ; 0
v1.8.2 ; 0
v1.8.3 ; 0
v1.8.4 ; 0
v1.9.0-0.1.rc1 ; 3
v1.9.0-0.2.rc2 ; 0
v1.9.0-1 ; 0
v1.9.0-2 ; 0
v1.9.0.1-1 ; 0
v1.9.1-0.1.rc1 ; 0
v1.9.1-0.2.rc2 ; 0
v1.9.1-0.3.rc3 ; 0
v1.9.1-0.4.rc4 ; 0
v1.9.1-1 ; 0
v1.9.2-0.1.rc1 ; 0
v1.9.2-0.2.rc2 ; 0
v1.9.2-1 ; 0
v1.9.3-0.1.rc1 ; 0
v1.9.3-0.2.rc2 ; 0
v1.9.3-0.3.rc3 ; 0
v1.9.3-1 ; 0
v1.9.4-0.1.rc1 ; 0
v1.9.4-0.2.rc2 ; 0
v1.9.4-0.3.rc3 ; 0
v1.9.4-1 ; 0
v1.9.5-0.1.rc1 ; 0
v1.9.5-1 ; 0
v1.9.6-0.1.rc1 ; 0
v1.9.6-1 ; 0
v2.0.0-0.6.rc1 ; 0
v2.0.0-0.7.rc2 ; 0
v2.0.0-0.8.rc3 ; 0
v2.0.0-0.9.rc4 ; 0
v2.0.0.0-1 ; 0
v2.0.0.1-1 ; 0
v2.0.0.2-1 ; 0
v2.0.1.0-0.1.rc1 ; 0
v2.0.1.0-0.2.rc2 ; 0
v2.0.1.0-1 ; 0
v2.0.2.0-0.1.rc1 ; 0
v2.0.2.0-0.2.rc2 ; 0
v2.0.2.0-0.3.rc3 ; 0
v2.0.2.0-0.4.rc4 ; 0
v2.0.2.0-1 ; 0
v2.1.0.0-0.1.rc1 ; 0
v2.1.0.0-0.2.rc2 ; 0
v2.1.0.0-0.3.rc3 ; 0
v2.1.0.0-0.4.rc4 ; 0
v2.1.0.0-1 ; 0
v2.1.1.0-0.1.rc1 ; 0
v2.1.1.0-0.2.rc2 ; 0
v2.1.1.0-0.3.rc3 ; 0
v2.1.1.0-0.4.rc4 ; 0
v2.1.1.0-0.5.rc5 ; 0
v2.1.1.0-1 ; 0
```

**Comment by ansibot** [comment]
From @jctanner on 2015-06-12T15:05:39Z

Based on the results above, I can not reproduce the described problem with any version of ansible. @tommarshall please review the testcase I gave and let me know if I've missed something.

needs_info

**Comment by ansibot** [comment]
From @willthames on 2015-06-12T15:05:39Z

@jctanner, without the full command line for the test, it's difficult to know why you can't reproduce with e.g. 2.1.1-0.0.2.rc2

Note that it only happens in checkmode.

**Comment by ansibot** [comment]
From @danschmidt5189 on 2015-06-12T15:05:39Z

@jctanner As @willthames mentioned, your test ought to fail in check mode. Here's a detailed rundown:

## ./site.yml

``` yaml
#! ansible-playbook -v
---

- hosts: 10.0.0.10
  gather_facts: no
  vars:
    checkoutdir: /tmp/ansible-modules-core
    repo: https://github.com/ansible/ansible-modules-core
    version: devel
  pre_tasks:
    - name: Ensure git installed.
      yum: pkg=git state=present update_cache=yes
      become: yes
  tasks:
    - name: Clone the repo, twice.
      git: repo={{ repo }}
           dest={{ checkoutdir }}
           version={{ version }}
           force=yes
      with_sequence: count=2
      register: clones
    - name: Attack of the clones!
      debug: var=clones
    - name: Ensure the second clone did not result in changes.
      assert:
        that:
          - not clones.results[1].changed

...
```

## ./Vagrantfile

``` ruby
Vagrant.configure("2") do |config|
  config.vm.box = "geerlingguy/centos6"
  config.vm.network :private_network, ip: '10.0.0.10'
end
```

## ./ansible.cfg

``` ini
[defaults]
host_key_checking=false
private_key_file=.vagrant/machines/default/virtualbox/private_key
remote_user=vagrant
```

## ./git-check-bug.sh

```
#! /bin/sh
vagrant destroy -f && vagrant up
ansible-playbook site.yml -i 10.0.0.10, # passes
ansible-playbook site.yml -i 10.0.0.10, -CD # fails
```

**Comment by ansibot** [comment]
From @jctanner on 2015-06-12T15:05:39Z

@willthames and @danschmidt5189 my test command line does not contain --check ...

```
ansible-playbook -vvvv -i inventory site.yml
```

I've gone back over this issue a few times now, and I don't see any indication from the submitter that check mode was used. Check mode was only part of the discussion after @nirik's comment. @tommarshall's description only makes sense if the playbook is not in checkmode, because the directory would not be moved prior to the final git task ... aka, his playbook doesn't work in checkmode.

needs_info ... from the submitter.

**Comment by ansibot** [comment]
From @danschmidt5189 on 2015-06-12T15:05:39Z

K, I'll wait on a response from @tommarshall before opening a new issue for the check mode bug.

To clarify, as it stands we have a reproducible test case indicating the git module returns "changed" when run in check mode, even when there would be no changes. The issue is whether that was the actual problem faced by the submitter; in any case, it _is_ a problem.

**Comment by ansibot** [comment]
From @willthames on 2015-06-12T15:05:39Z

@danschmidt5189 the check mode bug is fixed in Ansible 2.1.1, afaik (see #3025)

**Comment by ansibot** [comment]
From @danschmidt5189 on 2015-06-12T15:05:39Z

@willthames FWIW I ran my tests against 2.2.0:

```
$ ansible --version
ansible 2.2.0 (detached HEAD 134d70e7b9) last updated 2016/07/15 16:43:54 (GMT -700)
  lib/ansible/modules/core: (detached HEAD 7de287237f) last updated 2016/07/15 16:43:58 (GMT -700)
  lib/ansible/modules/extras: (detached HEAD 68ca157f3b) last updated 2016/07/15 16:44:01 (GMT -700)
  config file = /path/to/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

**Comment by ansibot** [comment]
From @willthames on 2015-06-12T15:05:39Z

@danschmidt5189 ok, that is a different issue to #3025, which only applies to `HEAD`, not a named branch.

**Comment by ansibot** [comment]
From @tommarshall on 2015-06-12T15:05:39Z

Check mode was not used, however the issue is still present.

I've been able to further reduce the test case to the following:

#### `site.yml`

``` yml
- hosts: 192.168.33.44
  gather_facts: no

  pre_tasks:
    - name: Ensure git installed.
      apt: pkg=git state=present update_cache=yes
      become: yes

  tasks:
    - name: Clone the scripts repository 1.
      git:
        repo: ssh://git@repo.example.com:/example.com/server-scripts.git
        dest: /tmp/scripts
        accept_hostkey: true

    - name: Clone the scripts repository 2.
      git:
        repo: ssh://git@repo.example.com:/example.com/server-scripts.git
        dest: /tmp/scripts
        accept_hostkey: true
      register: second_clone

    - debug: var=second_clone
```

#### `Vagrantfile`

``` rb
Vagrant.configure("2") do |config|
  config.vm.box = "ubuntu/trusty64"
  config.vm.box_url = "https://vagrantcloud.com/ubuntu/trusty64/version/1/provider/virtualbox.box"
  config.vm.network :private_network, ip: '192.168.33.44'
end
```

#### `ansible.cfg`

``` ini
[defaults]
host_key_checking=false
private_key_file=.vagrant/machines/default/virtualbox/private_key
remote_user=vagrant
```

#### `inventory`

``` ini
[all]
192.168.33.44
```

#### `run.sh`

``` sh
#! /bin/sh
vagrant destroy -f > /dev/null && vagrant up > /dev/null
ansible-playbook -vv site.yml -i inventory
ansible --version
```

I can confirm this issue is present on `1.9.6`:

```
$ ./run.sh

PLAY [192.168.33.44] **********************************************************

TASK: [Ensure git installed.] *************************************************
...

TASK: [Clone the scripts repository 1.] ***************************************
<192.168.33.44> REMOTE_MODULE git dest=/tmp/scripts repo=ssh:********@repo.example.com:/example.com/server-scripts.git
changed: [192.168.33.44] => {"after": "e8fc45f1fe693426e25e1696b49d158475753414", "before": null, "changed": true}

TASK: [Clone the scripts repository 2.] ***************************************
<192.168.33.44> REMOTE_MODULE git dest=/tmp/scripts repo=ssh:********@repo.example.com:/example.com/server-scripts.git
changed: [192.168.33.44] => {"after": "e8fc45f1fe693426e25e1696b49d158475753414", "before": "e8fc45f1fe693426e25e1696b49d158475753414", "changed": true}

TASK: [debug var=second_clone] ************************************************
ok: [192.168.33.44] => {
    "var": {
        "second_clone": {
            "after": "e8fc45f1fe693426e25e1696b49d158475753414",
            "before": "e8fc45f1fe693426e25e1696b49d158475753414",
            "changed": true,
            "invocation": {
                "module_args": "",
                "module_complex_args": {
                    "accept_hostkey": true,
                    "dest": "/tmp/scripts",
                    "repo": "ssh://git@repo.example.com:/example.com/server-scripts.git"
                },
                "module_name": "git"
            }
        }
    }
}

PLAY RECAP ********************************************************************
192.168.33.44              : ok=4    changed=3    unreachable=0    failed=0

ansible 1.9.6
  configured module search path = None
```

and `2.1.1.0`:

```
$ ./run.sh
Using /Users/tom/tmp/ansible-git-bug-1509/ansible.cfg as config file

PLAYBOOK: site.yml *************************************************************
1 plays in site.yml

PLAY [192.168.33.44] ***********************************************************

TASK [Ensure git installed.] ***************************************************
task path: /Users/tom/tmp/ansible-git-bug-1509/site.yml:5
...

TASK [Clone the scripts repository 1.] *****************************************
task path: /Users/tom/tmp/ansible-git-bug-1509/site.yml:10
changed: [192.168.33.44] => {"after": "e8fc45f1fe693426e25e1696b49d158475753414", "before": null, "changed": true, "warnings": []}

TASK [Clone the scripts repository 2.] *****************************************
task path: /Users/tom/tmp/ansible-git-bug-1509/site.yml:16
changed: [192.168.33.44] => {"after": "e8fc45f1fe693426e25e1696b49d158475753414", "before": "e8fc45f1fe693426e25e1696b49d158475753414", "changed": true, "warnings": []}

TASK [debug] *******************************************************************
task path: /Users/tom/tmp/ansible-git-bug-1509/site.yml:23
ok: [192.168.33.44] => {
    "second_clone": {
        "after": "e8fc45f1fe693426e25e1696b49d158475753414",
        "before": "e8fc45f1fe693426e25e1696b49d158475753414",
        "changed": true,
        "warnings": []
    }
}

PLAY RECAP *********************************************************************
192.168.33.44              : ok=4    changed=3    unreachable=0    failed=0

ansible 2.1.1.0
  config file = /Users/tom/tmp/ansible-git-bug-1509/ansible.cfg
  configured module search path = Default w/o overrides
```

I've removed the git install output, and replaced the clone domain name with example.com. Other than that the output is unaltered.

It appears to be specific to the repository involved. Unfortunately that repository is private, and I haven't yet been able to find an occurance of this issue with a public repository.

As far as I'm aware there's nothing special about this repository.

It's in a clean state after the initial clone:

```
vagrant@vagrant-ubuntu-trusty-64:/tmp/scripts$ git status
On branch master
Your branch is up-to-date with 'origin/master'.

nothing to commit, working directory clean
```

The repository is under active development, and this issue is always present, so I don't believe it's associated to a particualr commit:

```
vagrant@vagrant-ubuntu-trusty-64:/tmp/scripts$ git log --pretty=format:"%h - %ar" | head -n5
e8fc45f - 4 weeks ago
7c20933 - 4 weeks ago
1b94dd0 - 5 weeks ago
c76b34c - 5 weeks ago
aefb035 - 8 weeks ago
```

The repo does contain submodules.

I appreciate this could be tough to track down without a public repo. Happy to answer any questions you have about the repo in question.

Thanks all for looking into this.

Tom

**Comment by ansibot** [comment]
From @tommarshall on 2015-06-12T15:05:39Z

I think the submodules could be at fault here. `.gitmodules` in the repository lists 6 submodules:

```
vagrant@vagrant-ubuntu-trusty-64:/tmp/scripts$ cat .gitmodules
[submodule "fs/var/www/linux-dash"]
    path = fs/var/www/linux-dash
    url = https://github.com/afaqurk/linux-dash
[submodule "fs/etc/nginx/sslconfig"]
    path = fs/etc/nginx/sslconfig
    url = https://github.com/cloudflare/sslconfig.git
[submodule "fs/etc/modsecurity/owasp-modsecurity-crs"]
    path = fs/etc/modsecurity/owasp-modsecurity-crs
    url = https://github.com/SpiderLabs/owasp-modsecurity-crs.git
[submodule "fs/etc/nagios/shellshocker"]
    path = fs/etc/nagios/shellshocker
    url = https://github.com/wreiske/shellshocker
[submodule "fs/etc/skel"]
    path = fs/etc/skel
    url = git@repo.example.com:example.com/dotfiles
[submodule "fs/opt/dotfiles"]
    path = fs/opt/dotfiles
    url = git@repo.example.com:example.com/dotfiles
```

But the `.git/config` only shows the first three:

```
vagrant@vagrant-ubuntu-trusty-64:/tmp/scripts$ cat .git/config
[core]
    repositoryformatversion = 0
    filemode = true
    bare = false
    logallrefupdates = true
[remote "origin"]
    url = ssh://git@repo.example.com:/example.com/server-scripts.git
    fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
    remote = origin
    merge = refs/heads/master
[submodule "fs/etc/modsecurity/owasp-modsecurity-crs"]
    url = https://github.com/SpiderLabs/owasp-modsecurity-crs.git
[submodule "fs/etc/nginx/sslconfig"]
    url = https://github.com/cloudflare/sslconfig.git
[submodule "fs/var/www/linux-dash"]
    url = https://github.com/afaqurk/linux-dash
```

**Comment by ansibot** [comment]
From @jctanner on 2015-06-12T15:05:39Z

@tommarshall what is in those configs before and after the second clone?

needs_info

**Comment by ansibot** [comment]
From @tommarshall on 2015-06-12T15:05:39Z

@jctanner The contents of `.gitmodules` & `.git/config` are as shown above before the second clone, and are unchanged after the second clone.

**Comment by ansibot** [comment]
From @tommarshall on 2015-06-12T15:05:39Z

So it turns out the 3 submodules present in `.gitmodules` but not `.git/config` are left over from an incomplete delete operation. Each of those 3 submodule paths had been deleted from the repo without deleting the lines from `.gitmodules`.

I've cleaned up the `.gitmodules` file and 'changed' is now being reported correctly.

The repo was clearly at fault, but that doesn't explain the false positive 'changed' status from the module.

I think the cause lies in `submodules_fetch`, specifically: https://github.com/ansible/ansible-modules-core/blob/devel/source_control/git.py#L579-L586

It appears to read each of the submodule paths from `.gitmodules` and sets changed `True` if the path does not exist.

That's one way to handle this edge case, but it appears `git` itself is more forgiving and drops/ignores any submodules defined in `.gitmodules` without a corresponding path when generating the `.git/config` file during a `clone`.

Thanks @jctanner and everyone else for your time on this.

**Comment by ansibot** [comment]
From @jctanner on 2015-06-12T15:05:39Z

@tommarshall 2 changes made this afternoon ...

https://github.com/ansible/ansible-modules-core/pull/3364
https://github.com/ansible/ansible-modules-core/pull/4392

With those, we should at least be able to understand in the future that the change bit has been caused by submodules.

**Comment by ansibot** [comment]
From @jctanner on 2015-06-12T15:05:39Z

@tommarshall was the "incomplete delete operation" from the git module or another ansible module on the local checkout? Or was it upstream in the remote?

needs_info

**Comment by ansibot** [comment]
From @tommarshall on 2015-06-12T15:05:39Z

@jctanner it was already present in the history from the remote.

**Comment by ansibot** [comment]
@ansibot Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @robinro
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by s-hertel** [comment]
https://github.com/ansible/ansible-modules-core/pull/4392 improved this by giving a clue as to what is changing. We still return the status 'changed', which we may be able to improve.

**Comment by nitzmahone** [comment]
We believe this has been improved in more recent versions- please open a new issue if you're still seeing this behavior in currently supported versions of Ansible.
